grudge/symbolic/operators.py

```diff
@@ -651,20 +658,31 @@ def matrix(self, afgrp, volgrp, dtype):
                     afgrp.nunit_dofs),
                 dtype=dtype)
 
-        from modepy.tools import UNIT_VERTICES
         import modepy as mp
-        for iface, fvi in enumerate(
-                volgrp.mesh_el_group.face_vertex_indices()):
-            face_vertices = UNIT_VERTICES[volgrp.dim][np.array(fvi)].T
-            matrix[:, iface, :] = mp.nodal_face_mass_matrix(
-                    volgrp.basis(), volgrp.unit_nodes, afgrp.unit_nodes,
-                    volgrp.order,
-                    face_vertices)
-
-        # np.set_printoptions(linewidth=200, precision=3)
-        # matrix[np.abs(matrix) < 1e-13] = 0
-        # print(matrix)
-        # 1/0
+
+        vol_basis = volgrp.basis_obj()
+        faces = mp.faces_for_shape(volgrp.shape)
+
+        for iface, face in enumerate(faces):
+            # NOTE: Quadrature degree determined such that
+            # we either guarantee exact integration of degree 2N
+            # polynomials, or use a higher-order quadrature rule
+            # provided by the user (for example, when doing overintegration)
+            face_quad_degree = max(2*volgrp.order, afgrp.order)
+            face_quadrature = mp.quadrature_for_space(
+                mp.space_for_shape(face, face_quad_degree),
+                face
+            )
+            assert face_quadrature.exact_to >= afgrp.order
+            face_basis = mp.basis_for_space(
+                mp.space_for_shape(face, afgrp.order),
+                face
+            )
```

Comment:
Why not ask the `afgrp` for its basis?

Reply:
Ah. Still though, if it *does* have one, IMO we should use it. `if isinstance(afgrp, ElementGroupWithBasis)`...

Reply:
Okay! Will add this in